drivers: Kconfig.nrfx: Choose driver type basing on DT compatible (for I2C, SPI, and UART drivers)

anangl opened the pull request:
Remove prompts from Kconfig options `SPI_x_NRF_SPI*` that select the
type of nrfx driver (for SPI, SPIM, or SPIS peripheral) to be used
for a given instance. This prevents the options from being modified
in configuration files.
Instead, make one of these options selected by default according to the
"compatible" property set for the corresponding SPI node in devicetree.

This eliminates the need of changing both the "compatible" property in
devicetree and the Kconfig option selecting the driver type when a user
wants to switch between SPI, SPIM, and SPIS for a given instance.

Since all `SPI_x_NRF_SPI*` options are made "hidden" by this commit,
all their occurrences in configuration files are removed.

Same changes are done to Kconfig options `I2C_x_NRF_TWI*` and `UART_x_NRF_UART*`.

Additionally, I2C and SPI drivers are made no longer dependent on Kconfig
options that enable instances (i.e. `I2C_x` and `SPI_x`). Now these drivers
enable hardware instances when corresponding nodes in devicetree are
enabled (have status "okay").
Also dependencies on `HAS_HW_NRF_*` options are removed from Kconfigs
of I2C, SPI, and UART drivers, as for hidden options that activate
proper type of driver such dependencies are not actually helpful.

anangl committed:
drivers/spi/Kconfig.nrfx: Choose driver type basing on DT compatible

Remove prompts from Kconfig options `SPI_x_NRF_SPI*` that select the
type of nrfx driver (for SPI, SPIM, or SPIS peripheral) to be used
for a given instance. This prevents the options from being modified
in configuration files.
Instead, make one of these options selected by default according to the
"compatible" property set for the corresponding SPI node in devicetree.

This eliminates the need of changing both the "compatible" property in
devicetree and the Kconfig option selecting the driver type when a user
wants to switch between SPI, SPIM, and SPIS for a given instance.

Since all `SPI_x_NRF_SPI*` options are made "hidden" by this commit,
all their occurrences in configuration files are removed.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
What I want to achieve is to eliminate the need of changing the related Kconfig option when the "compatible" property in a given dts SPI node is changed, let's say from `nordic,nrf-spi` to `nordic,nrf_spim` so that `SPIM` peripheral is used instead of `SPI`.
The problem is that the solution presented in this PR uses the `dt_int_val` function that has been deprecated. What can I use instead?

anangl commented:
I need to use a similar solution for `i2c` and `uart` nodes. I included only the commit for `spi` nodes to not create too much noise.

ulfalizer commented on drivers/spi/Kconfig.nrfx:
```diff
@@ -35,18 +35,17 @@ config SOC_NRF52832_ALLOW_SPIM_DESPITE_PAN_58
 config SPI_0_NRF_SPI
-	bool "nRF SPI 0"
+	def_bool y
+	depends on $(dt_int_val,DT_NORDIC_NRF_SPI_SPI_0_BASE_ADDRESS) != 0
```
If you want to simplify it some more, you could do

```
def_bool $(dt_int_val,DT_NORDIC_NRF_SPI_SPI_0_BASE_ADDRESS) != 0
```

and get rid of the `depends on` (and ditto for the other ones).

Any kind of expression can appear after `default`, `def_bool`, etc.

anangl replied:
Will do. Thanks for the hint.

ulfalizer reviewed (approved):
Could consider turning those two-option choices into a plain `bool` symbol too, if they'd only ever have two options in practice.

ulfalizer commented:
> The problem is that the solution presented in this PR uses the `dt_int_val` function that has been deprecated. What can I use instead?

Good question. @galak deprecated them because we want to get rid of the `generated_dts_board.conf` file, and those functions reference identifiers in that one.

Will need some other way to refer to those property values if we remove those functions. Not sure what it'd look like off the top of my head though.

anangl commented:
Applied @ulfalizer's suggestion regarding simplifying `def_bool` definitions.

pabigot commented:
I think this makes some sense, but it seems to lack clear instructions on how a developer selects a non-default driver.  We certainly don't want people to have to modify the base board devicetree files.

An approach that makes sense to me is to have a set of overlay files like `nordic-spi1-spim.dtsi`:

    &spi1 {
        compatible = "nordic,nrf-spim";
        status = "okay";
    };

that would be `#include`d from an application-specific overlay.

If that's not the proposed solution, what is?

galak reviewed (dismissed):
dt_int_val is deprecated, please look at scripts/kconfig/kconfigfunctions.py if one of the other functions works or if we need to introduce a new function.

anangl commented:
@galak 
> dt_int_val is deprecated, please look at scripts/kconfig/kconfigfunctions.py if one of the other functions works or if we need to introduce a new function.

I know it's deprecated and I mentioned it in my first [comment](https://github.com/zephyrproject-rtos/zephyr/pull/19480#issuecomment-536549382) (now I see I should have put it as a part of the PR description, to make it more visible). I looked at `kconfigfunctions.py` but haven't found anything suitable, that's why I made this PR a draft one and labeled it as RFC. I think that a new function will need to be introduced, but I wanted to get some opinions from people more involved in dts processing scripts, hence this PR.

anangl commented:
@pabigot 
> An approach that makes sense to me is to have a set of overlay files like `nordic-spi1-spim.dtsi`:
> 
> ```
> &spi1 {
>     compatible = "nordic,nrf-spim";
>     status = "okay";
> };
> ```
> 
> that would be `#include`d from an application-specific overlay.

Yes, that can be a next step. First I wanted to make it possible to select the driver without an additional change in Kconfig. However, I'm not sure if including such file from the application overlay would be actually easier or more convenient for users than directly modifying the `compatible` property (in the same overlay file - I never thought of forcing people to modify the base board devicetree files).

pabigot commented:
> However, I'm not sure if including such file from the application overlay would be actually easier or more convenient for users than directly modifying the `compatible` property

I'm not sure either, but since you have to in theory change both `compatible` and `status` it seems more clear to recommend:
```
#include <nordic_i2c0_twim.dtsi>
#include <nordic_spi1_spis.dtsi>
```
over explicit node overlays to express a desire for I2C0 in its TWIM personality and SPI1 in its SPIS personality.  But it's certainly not necessary.

This is the approach taken for the Particle boards where the header SPI pins can be associated with different peripherals (you'd use `#include <mesh_feather_spi_spi3.dtsi>` to select SPI3), though it doesn't override which personality the peripheral uses.

anangl commented:
Rebased on top of #22612, to be able to utilize `node.labels`.

anangl committed:
scripts: kconfigfunctions: Add dt_nodelabel_has_compat function

This function takes a 'label' and returns "y" if an "enabled" node with
such label can be found in the EDT and that node is compatible with the
provided 'compat', otherwise it returns "n".

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
@galak I replaced uses of the deprecated `dt_int_val` function with the newly introduced  `dt_nodelabel_has_compat` (I'm not convinced about the name, but I couldn't come up with some more adequate one so far). Do you think such solution could be acceptable?

galak commented:
> @galak I replaced uses of the deprecated `dt_int_val` function with the newly introduced `dt_nodelabel_has_compat` (I'm not convinced about the name, but I couldn't come up with some more adequate one so far). Do you think such solution could be acceptable?

We want to move way from the alias based defines and move towards `DT_INST`.  So I'd ask we see how we can make that change for the NRF drivers and then I think the Kconfig selection will end up being greatly simplified.

mbolivar-nordic commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -349,6 +349,22 @@ def dt_compat_enabled(kconf, _, compat):
+    if doc_mode or edt is None:
+        return "n"
+
+    for node in edt.nodes:
```
I believe it would be faster to look up the enabled nodes with this compatible using `edt.compat2enabled[compat]`. This will avoid looping over all the nodes in the devicetree each time this function is called.

Then each node can be checked as `if label in node.labels: return "y"`.

anangl replied:
Thanks for the hint. Improved as suggested.

mbolivar-nordic commented on samples/bluetooth/hci_spi/boards/nrf51_pca10028.conf:
```diff
@@ -1 +0,0 @@
-CONFIG_SPI_1_NRF_SPIS=y
```
I love that this can be removed entirely now. This is a positive sign.

carlescufi commented:
@mbolivar-nordic and @galak please review

pabigot reviewed (approved):
Good; this has been a noticeable pain ever since the `M`-variants were added/became default.

I've run into this more with I2C, so a companion patch would be nice.

anangl commented:
> I've run into this more with I2C, so a companion patch would be nice.

Sure. I'll provide such (for UART as well) as a follow-up PR when this one gets merged. Or I can also extend this one if that's preferred.

pabigot commented:
Either way works for me.

anangl committed:
drivers/i2c/Kconfig.nrfx: Choose driver type basing on DT compatible

Remove prompts from Kconfig options `I2C_x_NRF_TWI*` that select
the type of nrfx driver (for TWI or TWIM peripheral) to be used
for a given instance. This prevents the options from being modified
from configuration files.
Instead, make one of these options selected by default according to the
"compatible" property set for the corresponding I2C node in devicetree.

This eliminates the need of changing both the "compatible" property in
devicetree and the Kconfig option selecting the driver type when a user
wants to switch between TWI and TWIM for a given instance.

Since all `I2C_x_NRF_TWI*` options are made "hidden" by this commit,
all their occurrences in configuration files are removed.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers/serial/Kconfig.nrfx: Choose driver type basing on DT compatible

Remove prompts from Kconfig options `UART_x_NRF_UART*` that select
the type of nrfx driver (for UART or UARTE peripheral) to be used
for a given instance. This prevents the options from being modified
from configuration files.
Instead, make one of these options selected by default according to the
"compatible" property set for the corresponding UART node in devicetree.

This eliminates the need of changing both the "compatible" property in
devicetree and the Kconfig option selecting the driver type when a user
wants to switch between UART and UARTE for a given instance.

Since all `UART_x_NRF_UART*` options are made "hidden" by this commit,
all their occurrences in configuration files are removed.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Added two commits with similar changes for I2C and UART related Kconfig options.

anangl commented:
Rebased on master.

mbolivar-nordic commented on samples/net/gsm_modem/boards/reel_board.conf:
```diff
@@ -1,2 +1 @@
 # Reel board UART
-CONFIG_UART_1_NRF_UARTE=y
```
This file can simply be removed now, right?

mbolivar-nordic commented on tests/drivers/uart/uart_async_api/boards/nrf52840_pca10056.conf:
```diff
@@ -1 +0,0 @@
-CONFIG_UART_1_NRF_UARTE=y
```
This one too.

galak reviewed (changes requested):
For i2c the Kconfig instances should all go away and just have that in the driver.
For spi, i'd ask if it makes sense to move "SPI Port N over-read character" as a dts property.
for uart, I'd think about the same question, can the per port Kconfig's become dts properties?

anangl committed:
drivers: nrf: Remove no longer suitable dependencies on Kconfig options

Make I2C and SPI drivers for nRF SoCs no longer dependent on Kconfig
options that enable instances (i.e. I2C_x and SPI_x). Now these drivers
enable hardware instances when corresponding nodes in devicetree are
enabled (have status "okay").
For I2C, SPI, and UART drivers, instead of using Kconfig dependencies
to prevent enabling of hardware instances that cannot be used together
(e.g. SPIM1 and TWIM1), a file that signals invalid configurations with
build assertions is added to compilation.
Also dependencies on HAS_HW_NRF_* options are removed from Kconfigs
of I2C, SPI, and UART drivers, as for hidden options that activate
proper type of driver such dependencies are not actually helpful.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Rebased on master.

anangl commented:
Rebased on master.

anangl commented:
Rebased on master, once more.

anangl committed:
drivers/gpio/Kconfig.sx1509b: Enable driver automatically when needed

Enable the gpio_sx1509b driver by default when a "semtech,sx1509b"
compatible node is enabled in devicetree.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
tree-wide: Remove assignments to Kconfig options unused in nRF drivers

Kconfig options that enable I2C and SPI instances are no longer used
in nRF drivers. Remove all assignments done to these options in related
board definitions, samples, and tests.

For nrf52_pca20020, also no longer needed setting of default values
for GPIO_SX1509B* options is removed (now the gpio_sx1509b driver is
enabled by default when a corresponding devicetree node is enabled).

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
> For i2c the Kconfig instances should all go away and just have that in the driver.
> For spi, i'd ask if it makes sense to move "SPI Port N over-read character" as a dts property.
> for uart, I'd think about the same question, can the per port Kconfig's become dts properties?

@galak I removed dependencies on Kconfig instances from I2C and SPI drivers. The other changes you suggest (moving stuff from Kconfig to dts properties) I will address in another PR (this one is already big enough and those are not that much related to this PR's topic).
Could you take another look, please?

mbolivar-nordic commented on boards/arm/nrf52833_pca10100/doc/index.rst:
```diff
@@ -178,11 +178,7 @@ more than one UART for connecting peripheral devices:
-   (top menu) -> Device Drivers -> Serial Drivers -> nRF UART nrfx drivers
-
-3. Use the UART1 as ``device_get_binding("UART_1")``
+2. Use the UART1 as ``device_get_binding("UART_1")``
```
```suggestion
2. Use the UART1 as ``device_get_binding(DT_LABEL(DT_NODELABEL(uart1)))``
```

Since you're here, if you have time. No worries if not.

mbolivar-nordic commented on boards/arm/nrf52840dk_nrf52840/doc/index.rst:
```diff
@@ -189,11 +189,7 @@ more than one UART for connecting peripheral devices:
-   (top menu) -> Device Drivers -> Serial Drivers -> nRF UART nrfx drivers
-
-3. Use the UART1 as ``device_get_binding("UART_1")``
+2. Use the UART1 as ``device_get_binding("UART_1")``
```
```suggestion
2. Use the UART1 as ``device_get_binding(DT_LABEL(DT_NODELABEL(uart1)))``
```

mbolivar-nordic commented on soc/arm/nordic_nrf/validate_enabled_instances.c:
```diff
@@ -0,0 +1,48 @@
+	"SPI"#idx", SPIM"#idx", SPIS"#idx", TWI"#idx", TWIM"#idx", TWIS"#idx \
+	IF_ENABLED(CONFIG_SOC_SERIES_NRF53X, (", UARTE"#idx)) \
+	IF_ENABLED(CONFIG_SOC_SERIES_NRF91X, (", UARTE"#idx)) \
+	". Check nodes with status \"okay\" in devicetree."
```
```suggestion
	". Check nodes with status \"okay\" in zephyr.dts."
```

The file is referred to as zephyr.dts in the documentation, so this may be easier to find.

mbolivar-nordic reviewed (approved):
Very nice. Some non-blocking minor ideas.

galak reviewed (commented):
Why do you need CONFIG_ I2C_x_NRF_TWI or CONFIG_I2C_x_NRF_TWIM at all?  Seems like you can do all that in the driver without the Kconfig symbols.

anangl commented:
> Why do you need CONFIG_ I2C_x_NRF_TWI or CONFIG_I2C_x_NRF_TWIM at all? Seems like you can do all that in the driver without the Kconfig symbols.

Yes, in `i2c_nrfx_twi*` drivers, which are actually shims (they use nrfx drivers underneath), I could enable instances without using any Kconfig symbols. But I also need to enable the corresponding driver and instance in nrfx (i.e. do `select NRFX_TWIx` or `select NRFX_TWIMx`) for these shims to work. And that's why I decided to keep `I2C_x_NRF_TWIM*` options (as hidden now) and still refer these options from `i2c_nrfx_twi*`. An alternative would be to move calls to `dt_nodelabel_has_compat` functions directly to `NRFX_TWI*` options and add dependency on `I2C_NRFX`. I thought that the way it is done now would be clearer. And users trying to figure out what happened to  `I2C_x_NRF_TWIM*` options (they were set from configuration files, now they can no longer be) still can grep for them.

anangl commented:
@mbolivar-nordic I applied your suggestions. Thanks.

galak commented:
> @galak Yes, in `i2c_nrfx_twi*` drivers, which are actually shims (they use nrfx drivers underneath), I could enable instances without using any Kconfig symbols. But I also need to enable the corresponding driver and instance in nrfx (i.e. do `select NRFX_TWIx` or `select NRFX_TWIMx`) for these shims to work. And that's why I decided to keep `I2C_x_NRF_TWIM*` options (as hidden now) and still refer these options from `i2c_nrfx_twi*`. An alternative would be to move calls to the `dt_nodelabel_has_compat` function directly to `NRFX_TWI*` options and add dependency on `I2C_NRFX`. I thought that the way it is done now would be clearer. And users trying to figure out what happened to `I2C_x_NRF_TWIM*` options (they were set from configuration files, now they can no longer be) still can grep for them.

I'm ok to let this in for now, but we really need to remove per instance Kconfig symbols.

anangl commented:
Rebased on master, to get the hotfix from #24017.

anangl commented:
Rebased on master, to get the fix from #24024.

anangl commented:
@carlescufi @galak This one is ready to be merged, provided that the last [DNM] commit is removed first.

carlescufi commented:
Merging since CI passed with the shippable hack

anangl commented:
Removed the [DNM] commit.
